### DISCUSSION TITLE
How to set up ccache to work with another compiler wrapper?

### DISCUSSION THREAD

**Discussion by haiomaio2** [ccache/ccache#426] [Questions & answers] [answered]
### Question ###

What is the reason for having ccache -s always 0 ? 

I followed this
```
cp ccache /usr/local/bin/
ln -s ccache /usr/local/bin/gcc
ln -s ccache /usr/local/bin/g++
ln -s ccache /usr/local/bin/cc
ln -s ccache /usr/local/bin/c++
```
Tried also this
`export PATH=/usr/lib/ccache:$PATH`

But nothing helped.

make
ccache -s
```
cache hit (direct)                     0
cache hit (preprocessed)               0
cache miss                             0
cache hit rate                      0.00 %
cleanups performed                     0
files in cache                         0
cache size                           0.0 kB
max cache size                       5.0 GB
```

In my ccache log I read only
```
[2019-05-17T13:43:15.127190 22675] === CCACHE 3.7.1+23_g6821579 STARTED =========================================
[2019-05-17T13:46:03.620166 22802] === CCACHE 3.7.1+23_g6821579 STARTED =========================================
[2019-05-17T13:46:47.651415 24233] === CCACHE 3.7.1+23_g6821579 STARTED =========================================
[2019-05-17T13:49:29.440854 27108] === CCACHE 3.7.1+23_g6821579 STARTED =========================================
```

**Comment by jrosdahl** [comment]
> export PATH=/usr/lib/ccache:$PATH

That should be `export PATH=/usr/local/bin:$PATH` if you installed ccache in `/usr/local/bin`.

> In my ccache log I read only [...] CCACHE [...] STARTED

Those lines were logged when you ran `ccache -s`, so that's not surprising.

As I already tried to explain in https://www.mail-archive.com/ccache@lists.samba.org/msg01458.html, you need to describe in more detail what you're trying to build, otherwise it's impossible to help you.

**Comment by haiomaio2** [comment]
The context is **cmake cross-compiling** with a **toolchain wrapper**.

```
$ export PATH=/usr/local/bin:$PATH

$ which ccache
/usr/local/bin/ccache
$ whereis ccache
ccache: /usr/bin/ccache /usr/lib/ccache /usr/local/bin/ccache /usr/share/man/man1/ccache.1.gz

$ which gcc
/usr/local/bin/gcc

$ export CCACHE_NOHASHDIR=true
```
What kind of information could I provide more? Is there a possibility to find out what at all is hashed (why the hits are always 0 ) ?

Further, I tried setting following
```
$ cd /usr/local/bin
$ ln -s ccache ext-toolchain-wrapper
```
It didnt work.
I tried also
```
$ ln -s ccache arm-am3354-linux-gnueabihf-gcc
$ ln -s ccache arm-am3354-linux-gnueabihf-g++
```
I even tried w/o success
`$ export CCACHE_PREFIX=/opt/sdk-dat-arm/host/usr/bin/ext-toolchain-wrapper`

It might be that ccache and cross-compiling with a wrapper are two incompatible worlds at the moment?

**Comment by afbjorklund** [comment]
If you have a modern cmake (i.e. 3.4+), the easiest is using CMAKE_C_COMPILER_LAUNCHER and CMAKE_CXX_COMPILER_LAUNCHER to do the wrapping rather than trying to change the compilers:

https://cmake.org/cmake/help/v3.14/prop_tgt/LANG_COMPILER_LAUNCHER.html

**Comment by haiomaio2** [comment]
Would it work with cross-compiling?

**Comment by afbjorklund** [comment]
Sure, it works with the normal CMAKE_C_COMPILER and CMAKE_CXX_COMPILER

**Comment by jrosdahl** [comment]
> Is there a possibility to find out what at all is hashed (why the hits are always 0 ) ?

It's because ccache is never invoked by what you are trying to build.

> The context is cmake cross-compiling with a toolchain wrapper.

Right, that's exactly what was needed to be able to help you.

As @afbjorklund already mentioned, use `CMAKE_C_COMPILER_LAUNCHER` and/or `CMAKE_CXX_COMPILER_LAUNCHER` if possible. Another alternative is something like this:

```
CC="ccache /toolchain/host/usr/bin/arm-am3354-linux-gnueabihf-gcc" \
CXX="ccache /toolchain/host/usr/bin/arm-am3354-linux-gnueabihf-g++" \
cmake ...
```

**Comment by haiomaio2** [comment]
I followed the advices above, however it is still to mention, that the problem is how to make ccache use the toolchain wrapper to which all the compilers are soft linked ....

```
$ sdk-dat-arm$ ll host/usr/bin/
arm-am3354-linux-gnueabihf-c++ -> ext-toolchain-wrapper*
arm-am3354-linux-gnueabihf-cc -> ext-toolchain-wrapper*
arm-am3354-linux-gnueabihf-cpp -> ext-toolchain-wrapper*
arm-am3354-linux-gnueabihf-g++ -> ext-toolchain-wrapper*
arm-am3354-linux-gnueabihf-gcc -> ext-toolchain-wrapper*
arm-am3354-linux-gnueabihf-gcc-4.9.4 -> ext-toolchain-wrapper*
```

```
sdk-dat-arm$ ll toolchain/bin/
arm-am3354-linux-gnueabihf-cc -> arm-am3354-linux-gnueabihf-gcc*
```
My ccache configuration
```
/usr/local/bin$ ls -al
arm-am3354-linux-gnueabihf-g++ -> ccache
arm-am3354-linux-gnueabihf-gcc -> ccache
cc -> ccache
ccache
g++ -> ccache
gcc -> ccache
```

So, even with the proposed approaches, I get still the same result when calling cmake
```
-- The C compiler identification is GNU 4.9.4
-- The CXX compiler identification is GNU 4.9.4
-- Check for working C compiler: /toolchain/host/usr/bin/arm-am3354-linux-gnueabihf-gcc
-- Check for working C compiler: /toolchain/host/usr/bin/arm-am3354-linux-gnueabihf-gcc -- works
```

Expected was maybe

```
-- Check for working C compiler: /usr/local/bin/ccache
-- Check for working C compiler: /usr/local/bin/ccache -- works
```

when following the approach 
```
CC="ccache /toolchain/host/usr/bin/arm-am3354-linux-gnueabihf-gcc" \
CXX="ccache /toolchain/host/usr/bin/arm-am3354-linux-gnueabihf-g++" \
cmake ...
```
than `make`

```
-- Check for working C compiler: /usr/local/bin/cc -- works
...
-- Check for working CXX compiler: /usr/local/bin/g++ -- works
...
```
cmake doesnt use the `arm-am3354-linux-gnueabihf-*`, but `/usr/local/bin/cc` and `/usr/local/bin/g++`. i.e. the above is ignored...

**Comment by jrosdahl** [comment]
I'm not familiar with "ext-toolchain-wrapper". It's kind of hard to give advice when we get more and more small bits of information that changes the situation. Why can't you just tell which project you are trying to build and exactly which commands you're running? Sorry about being a bit blunt, but I'm tired of guessing. Have you searched the Internet for reports from other users who tried to use ccache to build whatever you're building?

It seems like ext-toolchain-wrapper is using the masquerading technique, so then it's not a good idea to also let ccache masquerade as the compiler – there can be only one. Also, you should be aware of the caveats noted in [_Using ccache with other compiler_ wrappers](https://ccache.dev/manual/latest.html#_using_ccache_with_other_compiler_wrappers) in the ccache manual.

If it works to call e.g. `/toolchain/host/usr/bin/arm-am3354-linux-gnueabihf-gcc` directly, I would expect that something like this should work:

```
cmake \
  -DCMAKE_C_COMPILER=/toolchain/host/usr/bin/arm-am3354-linux-gnueabihf-gcc \
  -DCMAKE_CXX_COMPILER=/toolchain/host/usr/bin/arm-am3354-linux-gnueabihf-g++ \
  -DCMAKE_C_COMPILER_LAUNCHER=ccache \
  -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
 ...
```

If not, I don't have any more ideas with the given information.

**Comment by haiomaio2** [comment]
Since Im learning everything at once cmake, toolchain, ccache, the new project, my questions are  evolving, and I apologize  for not being that clear, and I really appreciate your help !!
The project consists of
- internal sources 
- http://www.ti.com/tool/PROCESSOR-SDK-AM335X
- the above two are packed with https://crosstool-ng.github.io/docs/toolchain-usage/ which creates the  "ext-toolchain-wrapper"
- the cmake configuration is as above proposed.

It think, that the key to the solution is how to configure the soft links of ccache. My ccache configuration is

```
/usr/local/bin$ ls -al
arm-am3354-linux-gnueabihf-g++ -> ccache
arm-am3354-linux-gnueabihf-gcc -> ccache
cc -> ccache
ccache
g++ -> ccache
gcc -> ccache
```

With the above however, I cannot compile the sources ...

**Comment by jrosdahl** [comment]
OK. Try this:

1. Don't use any symlinks for ccache (remove them).
2. Put `/toolchain/host/usr/bin` in your PATH (you already have that, I guess).
3. `touch empty.c`
4. `ccache -s`
5. `ccache arm-am3354-linux-gnueabihf-gcc -c empty.c`
6. `ccache -s`

Did step 5 work and did any ccache statistics counter change in step 6 compared to step 4? If so, it works to call the compiler via ccache like in step 5, so then it's only a matter of convincing cmake to run the compiler that way. And that should be possible with `-DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache` with a new enough cmake version.

Also, you can see what commands cmake's Makefile runs by running `make VERBOSE=1`. Verify that the compilation command lines look like expected.

**Comment by haiomaio2** [comment, marked as answer]
Your make `VERBOSE=1` advice turned out to be priceless. I could finally achieve that `cmake` works with [ccache](url) and the `toolchain wrapper` .

The solution I found was to adjust the `PATH` variable. Before I changed it, it looked like:
`export PATH=/usr/local/bin:/opt/sdk-dat-arm/toolchain/bin:$PATH`
the `toolchain/bin` contains all the cross-compiler binaries. However, calling them directly with `ccache` was not enough to resolve (include) dependencies (I guess, the peculiarity of the wrapper). So, I changed the `PATH` variable to
`export PATH=/usr/local/bin:/opt/sdk-dat-arm/host/usr/bin/:$PATH`
The `host/usr/bin` (see above post) contains the cross-compiler as soft links to the wrapper. The latter helped obviously `cmake` to resolve the (include) dependencies.

The approach I followed:
- I didn't remove the symlinks, because I discovered thanks to the `VERBOSE=1` that using the symlinks in `host/usr/bin` to the wrapper was solving my compilation problem.
- I used `set(CMAKE_C_COMPILER_LAUNCHER ccache) `to be on the safe side that there are no cmake side effects (cmake is still black magic to me), but I guess `-DCMAKE_XXX_COMPILER_LAUNCHER=ccache` would have also worked

If you could imagine a better approach, I would be curious to know. 

Many thanks for your help and patience.

**Comment by jrosdahl** [comment]
Great that you got it to work.

>  I guess -DCMAKE_XXX_COMPILER_LAUNCHER=ccache would have also worked

`CMAKE_CXX_COMPILER_LAUNCHER` (note: `CXX`, not `XXX`) is the for the C++ compiler, so if the code you're building only uses the C compiler it works fine to only use `CMAKE_C_COMPILER_LAUNCHER`.

**Comment by afbjorklund** [comment]
This should probably go into the documentation, how to make ccache work with Make and CMake ?

> To use the first method, just make sure that *ccache* is in your *PATH*.

Just something simple like how to use `$(CCACHE)` and how to use `CMAKE_C_COMPILER_LAUNCHER`

Good make a good addition to the current section, which focuses on (the simpler) masquerading...

**Comment by jrosdahl** [comment]
> This should probably go into the documentation

Sounds like a good idea.
